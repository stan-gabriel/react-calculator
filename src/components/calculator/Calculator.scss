@import '../../scss/variables';

.Calculator {
  display: flex;
  flex-direction: column;
  //height: 20rem;
  //width: 17rem;
  text-align: center;
  margin: 5rem;
}

.buttons-wrapper {
  display: flex;
}
