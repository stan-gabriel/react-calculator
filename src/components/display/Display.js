import React from 'react';
import './Display.scss';

class Display extends React.Component {

	adjustText = function (textLength) {
		if (textLength < 8) {
			return 3
		} else {
			return 24 / textLength
		}
	};

	render() {
		let initialLength = `${this.props.result} `;
		return (
			<div className="Display" style={{fontSize: `${this.adjustText(initialLength.toString().length)}rem`}}>
				{this.props.result}
			</div>
		)
	};
}

export default Display
