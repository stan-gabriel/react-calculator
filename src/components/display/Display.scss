@import '../../scss/variables';

.Display {
  display: flex;
  justify-content: flex-end;
  align-items: flex-end;
  background: $base-color;
  height: 5rem;
  width: 14.4rem;
  font-weight: 200;
  color: #ffff;
  //font-size: 2rem;
  font-size: 3.5vh;
  padding: 0 0.4rem;
  border-top-right-radius: 5px;
  border-top-left-radius: 5px;

  white-space: pre-line;
  word-wrap: break-word;
}
