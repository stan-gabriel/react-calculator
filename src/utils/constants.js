import { VALUE, CLEAR_ALL, CHANGE_SIGN, CALCULATE_RESULT, OPERATION, CALCULATE_PERCENTAGE } from '../reducers/types'

export const operations = {
	add: (a, b) => {
		return parseFloat(a) + parseFloat(b)
	},
	subtraction: (a, b) => {
		return parseFloat(a) - parseFloat(b)
	},
	multiplication: (a, b) => {
		return parseFloat(a) * parseFloat(b)
	},
	division: (a, b) => {
		return parseFloat(a) / parseFloat(b)
	},
};

export const rows = [
	[
		{value: 'AC', colour: 'dark-btn', meaning: CLEAR_ALL},
		{value: '+/-', colour: 'dark-btn', meaning: CHANGE_SIGN},
		{value: '%', colour: 'dark-btn', meaning: CALCULATE_PERCENTAGE},
		{value: 'รท', colour: 'orange-btn', meaning: OPERATION, action: operations.division}
	],
	[
		{value: '7', colour: 'light-btn', meaning: VALUE},
		{value: '8', colour: 'light-btn', meaning: VALUE},
		{value: '9', colour: 'light-btn', meaning: VALUE},
		{value: 'x', colour: 'orange-btn', meaning: OPERATION, action: operations.multiplication}
	],
	[
		{value: '4', colour: 'light-btn', meaning: VALUE},
		{value: '5', colour: 'light-btn', meaning: VALUE},
		{value: '6', colour: 'light-btn', meaning: VALUE},
		{value: '-', colour: 'orange-btn', meaning: OPERATION, action: operations.subtraction}
	],
	[
		{value: '1', colour: 'light-btn', meaning: VALUE},
		{value: '2', colour: 'light-btn', meaning: VALUE},
		{value: '3', colour: 'light-btn', meaning: VALUE},
		{value: '+', colour: 'orange-btn', meaning: OPERATION, action: operations.add}
	],
	[
		{value: '0', colour: 'special-btn', meaning: VALUE},
		{value: '.', colour: 'light-btn', meaning: VALUE},
		{value: '=', colour: 'orange-btn', meaning: CALCULATE_RESULT}
	],
];


export const CalculatorOperations = {
	'/': (a, b) => {return parseFloat(a) / parseFloat(b)},
	'*': (a, b) => {return parseFloat(a) * parseFloat(b)},
	'+': (a, b) => {return parseFloat(a) + parseFloat(b)},
	'-': (a, b) => {return parseFloat(a) - parseFloat(b)},
};
