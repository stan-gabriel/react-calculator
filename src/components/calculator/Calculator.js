import React from 'react';
import { connect } from 'react-redux';
import Display from '../display/Display'
import Button from '../key-button/KeyButton';
import { rows, CalculatorOperations } from '../../utils/constants';
import * as actions from '../../reducers/actions';
import { CHANGE_SIGN, CLEAR_ALL, OPERATION, CALCULATE_RESULT, CALCULATE_PERCENTAGE } from '../../reducers/types';
import './Calculator.scss';


class Calculator extends React.Component {
	static defaultProps = {
		operation: '',
		addFirstValue: () => {
		}, addOperationType: () => {
		}, addSecondValue: () => {
		}, changeSign: () => {
		}, clearAll: () => {
		}, calculateResult: () => {
		}, calculatePercentage: () => {
		}
	};

	componentDidMount() {
		document.addEventListener('keydown', this.handleKeyDown)
	}

	componentWillUnmount() {
		document.removeEventListener('keydown', this.handleKeyDown)
	}

	performAction = (buttonData) => event => {
		const {operation, addFirstValue, addOperationType, addSecondValue, changeSign, clearAll, calculateResult, calculatePercentage} = this.props;

		if (buttonData.meaning === CHANGE_SIGN) {
			return changeSign();
		}
		if (buttonData.meaning === CLEAR_ALL) {
			return clearAll();
		}
		if (buttonData.meaning === OPERATION) {
			return addOperationType(buttonData.value, buttonData.action);
		}
		if (buttonData.meaning === CALCULATE_RESULT) {
			return calculateResult()
		}
		if (buttonData.meaning === CALCULATE_PERCENTAGE) {
			return calculatePercentage()
		}
		if (!operation) {
			addFirstValue(buttonData.value)
		} else {
			addSecondValue(buttonData.value);
		}
	};

	handleKeyDown = event => {
		const {operation, addFirstValue, addOperationType, addSecondValue, clearAll, calculateResult, calculatePercentage} = this.props;
		let {key} = event;

		if (key === 'Enter')
			calculateResult();
		if ((/\d/).test(key)) {
			// event.preventDefault();
			if (!operation) {
				addFirstValue(key)
			} else {
				addSecondValue(key);
			}
		} else if (key in CalculatorOperations) {
			// event.preventDefault();
			addOperationType(key, CalculatorOperations[key]);
		} else if (key === '.') {
			// event.preventDefault();
			if (!operation) {
				addFirstValue(key)
			} else {
				addSecondValue(key);
			}
		} else if (key === '%') {
			// event.preventDefault();
			calculatePercentage()
		} else if (key === 'Backspace') {
			// event.preventDefault();
			clearAll();
			// TODO implement 'clear last char'
		}
	};

	renderButton = (buttonData, index) => {
		return <Button key={index} colour={buttonData.colour} value={buttonData.value}
		               onClick={this.performAction(buttonData)}/>
	};

	renderRow = (row, index) => {
		return (
			<div className='buttons-wrapper' key={index}>
				{row.map(this.renderButton)}
			</div>
		)
	};

	render() {
		const {displayValue} = this.props;
		return (
			<div className="Calculator">
				<Display result={displayValue}/>

				{rows.map(this.renderRow)}
			</div>
		);
	}
}

export const PureCalculator = Calculator;

function mapStateToProps(state) {
	return {
		firstValue: state.app.firstValue,
		secondValue: state.app.secondValue,
		displayValue: state.app.displayValue,
		operation: state.app.operation,
	}
}

function mapDispatchToProps(dispatch) {
	return {
		addFirstValue: (firstValue) => dispatch(actions.addFirstValue(firstValue)),
		addSecondValue: (secondValue) => dispatch(actions.addSecondValue(secondValue)),
		addOperationType: (operationType, callback) => dispatch(actions.addOperationType(operationType, callback)),
		clearAll: () => dispatch(actions.clearAll()),
		changeSign: () => dispatch(actions.changeSign()),
		calculateResult: () => dispatch(actions.calculateResult()),
		calculatePercentage: () => dispatch(actions.calculatePercentage())
	}
}

export default connect(mapStateToProps, mapDispatchToProps)(Calculator);
