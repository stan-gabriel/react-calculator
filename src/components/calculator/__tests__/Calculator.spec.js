import React from 'react';
import { shallow } from 'enzyme';
import { PureCalculator } from '../Calculator';
import Display from '../../display/Display'
import {
	CALCULATE_PERCENTAGE,
	CALCULATE_RESULT,
	CHANGE_SIGN,
	CLEAR_ALL,
	OPERATION,
	VALUE
} from "../../../reducers/types";

import { CalculatorOperations } from '../../../utils/constants'

describe('Calculator component', () => {
	let wrapper;

	beforeEach(() => (wrapper = shallow(<PureCalculator/>)));
	test('should match the snapshot', () => expect(wrapper).toMatchSnapshot());

	test('instance perform action does the right thing', () => {
		const changeSignMock = jest.fn();
		const clearAllMock = jest.fn();
		const addOperationTypeMock = jest.fn();
		const calculateResultMock = jest.fn();
		const calculatePercentageMock = jest.fn();
		const addSecondValueMock = jest.fn();
		wrapper = shallow(<PureCalculator changeSign={changeSignMock} clearAll={clearAllMock}
		                                  addOperationType={addOperationTypeMock} calculateResult={calculateResultMock}
		                                  calculatePercentage={calculatePercentageMock}
		                                  addFirstValue={addSecondValueMock}/>);
		const instance = wrapper.instance();
		instance.performAction({meaning: CHANGE_SIGN})();
		expect(changeSignMock.mock.calls).toHaveLength(1);
		instance.performAction({meaning: CLEAR_ALL})();
		expect(clearAllMock.mock.calls).toHaveLength(1);
		instance.performAction({meaning: OPERATION, value: 1, action: '+'})();
		expect(addOperationTypeMock.mock.calls).toHaveLength(1);
		expect(addOperationTypeMock.mock.calls[0][0]).toBe(1);
		expect(addOperationTypeMock.mock.calls[0][1]).toBe('+');
		instance.performAction({meaning: CALCULATE_RESULT})();
		expect(calculateResultMock.mock.calls).toHaveLength(1);
		instance.performAction({meaning: CALCULATE_PERCENTAGE})();
		expect(calculatePercentageMock.mock.calls).toHaveLength(1);
		instance.performAction({meaning: VALUE})();
		expect(addSecondValueMock.mock.calls).toHaveLength(1);
	});

	test('instance handleKeyDown does the right thing', () => {
		const calculateResultMock = jest.fn();
		const addFirstValuetMock = jest.fn();
		const addOperationTypeMock = jest.fn();
		const calculatePercentageMock = jest.fn();
		const clearAllMock = jest.fn();
		wrapper = shallow(<PureCalculator calculateResult={calculateResultMock} addFirstValue={addFirstValuetMock}
		                                  addOperationType={addOperationTypeMock} calculatePercentage={calculatePercentageMock} clearAll={clearAllMock}/>);
		const instance = wrapper.instance();
		instance.handleKeyDown({key: 'Enter'});
		expect(calculateResultMock.mock.calls).toHaveLength(1);
		instance.handleKeyDown({key: '1'});
		expect(addFirstValuetMock.mock.calls).toHaveLength(1);
		instance.handleKeyDown({key:'+', action: CalculatorOperations['+']});
		expect(addFirstValuetMock.mock.calls).toHaveLength(1);
		expect(addOperationTypeMock.mock.calls[0][0]).toBe('+');
		expect(addOperationTypeMock.mock.calls[0][1]).toBe(CalculatorOperations['+']);
		instance.handleKeyDown({key:'%'});
		expect(calculatePercentageMock.mock.calls).toHaveLength(1);
		instance.handleKeyDown({key: 'Backspace'});
		expect(clearAllMock.mock.calls).toHaveLength(1);
	});

	test('renderButton function check', () => {
		const spy = jest.spyOn(wrapper.instance(), 'renderButton');
		expect(spy).toHaveBeenCalledTimes(0);
		wrapper.instance().forceUpdate();
		expect(spy).toHaveBeenCalledTimes(19);
	});

	test('renderRow function check', () => {
		const spy = jest.spyOn(wrapper.instance(), 'renderRow');
		expect(spy).toHaveBeenCalledTimes(0);
		wrapper.instance().forceUpdate();
		expect(spy).toHaveBeenCalledTimes(5);
	});

	test('should render a <div />', () => {
		expect(wrapper.find('div').length).toEqual(6);
	});

	test('should render the container', () => {
		expect(wrapper.find('.Calculator').length).toEqual(1);
	});

	test('should render the Display Component', () => {
		expect(wrapper.containsMatchingElement(<Display />)).toEqual(true);
	});
});

