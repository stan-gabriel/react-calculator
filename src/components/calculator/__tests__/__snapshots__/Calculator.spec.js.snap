// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator component should match the snapshot 1`] = `
<div
  className="Calculator"
>
  <Display />
  <div
    className="buttons-wrapper"
    key="0"
  >
    <KeyButton
      colour="dark-btn"
      key="0"
      onClick={[Function]}
      value="AC"
    />
    <KeyButton
      colour="dark-btn"
      key="1"
      onClick={[Function]}
      value="+/-"
    />
    <KeyButton
      colour="dark-btn"
      key="2"
      onClick={[Function]}
      value="%"
    />
    <KeyButton
      colour="orange-btn"
      key="3"
      onClick={[Function]}
      value="รท"
    />
  </div>
  <div
    className="buttons-wrapper"
    key="1"
  >
    <KeyButton
      colour="light-btn"
      key="0"
      onClick={[Function]}
      value="7"
    />
    <KeyButton
      colour="light-btn"
      key="1"
      onClick={[Function]}
      value="8"
    />
    <KeyButton
      colour="light-btn"
      key="2"
      onClick={[Function]}
      value="9"
    />
    <KeyButton
      colour="orange-btn"
      key="3"
      onClick={[Function]}
      value="x"
    />
  </div>
  <div
    className="buttons-wrapper"
    key="2"
  >
    <KeyButton
      colour="light-btn"
      key="0"
      onClick={[Function]}
      value="4"
    />
    <KeyButton
      colour="light-btn"
      key="1"
      onClick={[Function]}
      value="5"
    />
    <KeyButton
      colour="light-btn"
      key="2"
      onClick={[Function]}
      value="6"
    />
    <KeyButton
      colour="orange-btn"
      key="3"
      onClick={[Function]}
      value="-"
    />
  </div>
  <div
    className="buttons-wrapper"
    key="3"
  >
    <KeyButton
      colour="light-btn"
      key="0"
      onClick={[Function]}
      value="1"
    />
    <KeyButton
      colour="light-btn"
      key="1"
      onClick={[Function]}
      value="2"
    />
    <KeyButton
      colour="light-btn"
      key="2"
      onClick={[Function]}
      value="3"
    />
    <KeyButton
      colour="orange-btn"
      key="3"
      onClick={[Function]}
      value="+"
    />
  </div>
  <div
    className="buttons-wrapper"
    key="4"
  >
    <KeyButton
      colour="special-btn"
      key="0"
      onClick={[Function]}
      value="0"
    />
    <KeyButton
      colour="light-btn"
      key="1"
      onClick={[Function]}
      value="."
    />
    <KeyButton
      colour="orange-btn"
      key="2"
      onClick={[Function]}
      value="="
    />
  </div>
</div>
`;
